src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -0,0 +1,261 @@
+            } finally {
+                CloserUtil.close(underlyingIterator);
+                backgroundThread = null;
+                currentBlock = null;
+            }
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (backgroundThread == null) {
+            throw new IllegalStateException("iterator has been closed");
+        }
+        if (currentBlock.hasNext()) return true;
+        
+        // rethrow any exceptions raised on the background thread
+        // at the point the exception would have been encountered
+        // if we had performed synchronous iteration
+        Throwable t = currentBlock.getException();
+        if (t != null) {
+            if (t instanceof Error) {
+                throw (Error) t;
+            } else if (t instanceof RuntimeException) {
+                throw (RuntimeException) t;
+            } else {
+                throw new RuntimeException(t);
+            }
+        }
+        if (currentBlock.isEndOfStream()) return false;
+        try {
+            // try loading next block
+            currentBlock = buffer.take();
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Error reading from background thread", e);
+        }
+        return hasNext();
+    }
+
+    @Override
+    public T next() {
+        if (hasNext()) {
+            return currentBlock.next();
+        }
+        throw new NoSuchElementException("next");
+    }
+
+    /**
+     * Performs 1 buffer worth of read-ahead on the underlying iterator
+     * (background thread method) 
+     */
+    private RecordBlock readAhead() {
+        if (!underlyingIterator.hasNext()) return new RecordBlock();
+        final List<T> readAhead = new ArrayList<>(bufferSize);
+        try {
+            for (int i = 0; i < bufferSize && underlyingIterator.hasNext(); i++) {
+                if (Thread.currentThread().isInterrupted()) {
+                    // eager abort if we've been told to stop
+                    return new RecordBlock(readAhead, new InterruptedException());
+                }
+                readAhead.add(underlyingIterator.next());
+            }
+            return new RecordBlock(readAhead);
+        } catch (Throwable t) {
+            return new RecordBlock(readAhead, t);
+        }
+    }
+    /**
+     * Background thread run loop
+     * @throws InterruptedException 
+     */
+    private void backgroundRun() {
+        try {
+            RecordBlock block;
+            do {
+                block = readAhead();
+                if (block.getException() instanceof InterruptedException) {
+                    // stop thread immediately if we've been told to stop
+                    return;
+                }
+                buffer.put(block);
+            } while (!block.isEndOfStream());
+        } catch (InterruptedException e) {
+            // stop thread
```

Comment:
as above: what happens if this thread dies while the buffer is empty and the main thread is blocking on the `take` call?